cloud/cloud_env_impl.cc

```diff
@@ -1840,6 +1844,150 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
+  if (!st.ok()) {
+    return st;
+  }
+
+  // Fetch latest Manifest file corresponds to current cookie
+  // NOTE: the latest Manifest file could be different from local Manifest file.
+  // TODO(wei): should we try to avoid overwriting the local Manifest file (with same name/epoch)?
+  st = FetchCurrentManifest(local_dbname);
+  if (!st.ok()) {
+    return st;
+  }
+
+  auto old_epoch = cloud_manifest->GetCurrentEpoch();
+
+  // Get the max file number of latest Manifest file.
+  // NOTE: the max file number could be different from VersionSet::next_file_num
+  uint64_t max_file_num;
+  st = ManifestReader::GetMaxFileNumberFromManifest(
+      GetBaseEnv(), ManifestFileWithEpoch(local_dbname, old_epoch),
+      &max_file_num);
+  if (!st.ok()) return st;
+
+  // rolling epoch of the Manifest file
+  auto new_epoch = generateNewEpochId();
+  cloud_manifest->AddEpoch(max_file_num, new_epoch);
+  const auto& fs = GetBaseEnv()->GetFileSystem();
+  st = CopyFile(fs.get(), ManifestFileWithEpoch(local_dbname, old_epoch),
+                ManifestFileWithEpoch(local_dbname, new_epoch), 0 /* size */,
+                true /* use_fsync */, nullptr /* io_tracer */,
+                Temperature::kUnknown);
+  if (!st.ok()) {
+    return st;
+  }
+
+  // Dump cloud_manifest into the CLOUDMANIFEST-new_cookie file
+  st = writeCloudManifest(cloud_manifest.get(),
+                          MakeCloudManifestFile(local_dbname, new_cookie));
+  if (!st.ok()) {
+    return st;
+  }
+
+  // upload new manifest
+  st = GetStorageProvider()->PutCloudObject(
+      ManifestFileWithEpoch(local_dbname, new_epoch), GetDestBucketName(),
+      ManifestFileWithEpoch(GetDestObjectPath(), new_epoch));
+  if (!st.ok()) {
+    return st;
+  }
+
+  // upload new cloud manifest
+  st = GetStorageProvider()->PutCloudObject(
+      MakeCloudManifestFile(local_dbname, new_cookie), GetDestBucketName(),
+      MakeCloudManifestFile(GetDestObjectPath(), new_cookie));
+
+  *existing_cookie_out = current_cookie;
+  return st;
+}
+
+Status CloudEnvImpl::SwitchCloudManifest(const std::string& local_dbname,
+                                         const std::string& new_cookie) {
+  std::string current_cookie = GetCookie();
+  if (new_cookie == current_cookie) {
+    return Status::InvalidArgument(
+        "New cloud manifest to switch to should have different cookie");
+  }
+
+  auto st = FetchCloudManifest(local_dbname, new_cookie);
+  if (!st.ok()) return st;
+
+  auto new_cloud_manifest_file_name =
+      MakeCloudManifestFile(local_dbname, new_cookie);
+  std::unique_ptr<SequentialFileReader> log;
+  st = SequentialFileReader::Create(GetBaseEnv()->GetFileSystem(),
+                                    new_cloud_manifest_file_name, FileOptions(),
+                                    &log, nullptr);
+  if (!st.ok()) return st;
+
+  st = cloud_manifest_->ReloadFromLog(std::move(log));
+  if (!st.ok()) return st;
+
+  SetCookie(new_cookie);
+  return st;
+}
```

Comment:
I don't think we should ever need to switch cloud manifests by downloading them from S3 and reloading the state from S3.

How about this:
* On startup, download the CLOUDMANIFEST with the cookie provided.
* If you are the leader, roll the cloud manifest by writing the delta to the local CLOUDMANIFEST file, write the delta to the replication log and also upload the CLOUDMANIFEST to S3 with the new cookie.
* If you are the follower, apply the delta you get from the replication log to the local CLOUDMANIFEST

 You only need to provide two new functions:
 1. RollCloudManifest(delta)
 2. UploadCloudManifestToCloud(new_cookie)
 
 Here delta is (max_file_number, epoch).
 
You call (1) on both the leader and the follower. (2) is only applicable to the leader.

Reply:
Also, are we going to call `RollCloudManifest(delta)` before or after CAS(cookie)?

I think either way has some issues:

- If we `CAS(cookie)` before `RollCloudManifest(delta)`, then the cloud manifest file is not ready in s3 yet after `CAS(cookie)`
- If we `CAS(cookie)` after `RollCloudManifest(delta)`, it's possible that `CAS(cookie)` fails, but the `CloudManifest` content has been changed on this follower. In the future, this follower will always contain an extra `pastEpoch`

I'm thinking about this:
```
// On Leader
RollCloudManifest(new_cookie, &delta) // fetch latest manifest file, upload CloudManifest and new Manifest to s3, return delta as serialized string
CAS(cookie)
ApplyCloudManifestDelta(delta) // apply delta to cloud manifest
Replicate delta to followers

// On Follower
ApplyCloudManifestDelta(delta) // apply delta to cloud manifest
```

Reply:
> But shouldn't we need to download the latest MANIFEST-epoch from s3 (on leader) to figure out the max_file_number? The local manifest file could be different from the one storing in s3. And I believe it might be better to do this in CloudEnv.

We will only roll the cloud manifest after opening the database, right? The process of opening the database itself will download the latest MANIFEST-epoch and read it. We will never need to re-download MANIFEST on a running database.

> Also, are we going to call RollCloudManifest(delta) before or after CAS(cookie)?

We need to do Roll before CAS, because we need CLOUDMANIFEST-cookie to exist before switching the cookie, as you point out.

> If we CAS(cookie) after RollCloudManifest(delta), it's possible that CAS(cookie) fails, but the CloudManifest content has been changed on this follower. In the future, this follower will always contain an extra pastEpoch. This can be a big issue, since on this follower, the same file number can be mapped to multiple epochs in CloudManifest.

There are two scenarios when CAS(cookie) fails:
1/ The new leader recovered the NewEpochRecord that contains the delta to the CLOUDMANIFEST from the old leader. In that case the leader and the follower should have exactly the same CLOUDMANIFEST contents.
2/ The new leader didn't recover the NewEpochRecord that the follower has applied to the local CLOUDMANIFEST. This is a diverged branch and the follower will have to resync from the new leader.

Reply:
> We will only roll the cloud manifest after opening the database, right? 
Wouldn't the follower need to roll cloud manifest whenever it tries to become leader? (and we want to avoid reopening the db here). 

> There are two scenarios when CAS(cookie) fails:
I don't fully get this part. IMO, CAS(cookie) can fail if there are two followers trying to become leader simultaneously. Only one of them should succeed when CAS(cookie)